File: rb/MonsterHunterWorld_sdk.rb
# MonsterHunterWorld SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'MonsterHunterWorld_types'


class MonsterHunterWorldSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = MonsterHunterWorldUtility.new
    @_utility = utility

    config = MonsterHunterWorldConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = MonsterHunterWorldHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = MonsterHunterWorldHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, MonsterHunterWorldFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    MonsterHunterWorldUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = MonsterHunterWorldHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = MonsterHunterWorldHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = MonsterHunterWorldHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = MonsterHunterWorldSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue MonsterHunterWorldError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = MonsterHunterWorldHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = MonsterHunterWorldHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.ailment.list / client.ailment.load({ "id" => ... })
  def ailment
    require_relative 'entity/ailment_entity'
    @ailment ||= AilmentEntity.new(self, nil)
  end

  # Deprecated: use client.ailment instead.
  def Ailment(data = nil)
    require_relative 'entity/ailment_entity'
    AilmentEntity.new(self, data)
  end


  # Idiomatic facade: client.armor.list / client.armor.load({ "id" => ... })
  def armor
    require_relative 'entity/armor_entity'
    @armor ||= ArmorEntity.new(self, nil)
  end

  # Deprecated: use client.armor instead.
  def Armor(data = nil)
    require_relative 'entity/armor_entity'
    ArmorEntity.new(self, data)
  end


  # Idiomatic facade: client.armor_set.list / client.armor_set.load({ "id" => ... })
  def armor_set
    require_relative 'entity/armor_set_entity'
    @armor_set ||= ArmorSetEntity.new(self, nil)
  end

  # Deprecated: use client.armor_set instead.
  def ArmorSet(data = nil)
    require_relative 'entity/armor_set_entity'
    ArmorSetEntity.new(self, data)
  end


  # Idiomatic facade: client.charm.list / client.charm.load({ "id" => ... })
  def charm
    require_relative 'entity/charm_entity'
    @charm ||= CharmEntity.new(self, nil)
  end

  # Deprecated: use client.charm instead.
  def Charm(data = nil)
    require_relative 'entity/charm_entity'
    CharmEntity.new(self, data)
  end


  # Idiomatic facade: client.decoration.list / client.decoration.load({ "id" => ... })
  def decoration
    require_relative 'entity/decoration_entity'
    @decoration ||= DecorationEntity.new(self, nil)
  end

  # Deprecated: use client.decoration instead.
  def Decoration(data = nil)
    require_relative 'entity/decoration_entity'
    DecorationEntity.new(self, data)
  end


  # Idiomatic facade: client.event.list / client.event.load({ "id" => ... })
  def event
    require_relative 'entity/event_entity'
    @event ||= EventEntity.new(self, nil)
  end

  # Deprecated: use client.event instead.
  def Event(data = nil)
    require_relative 'entity/event_entity'
    EventEntity.new(self, data)
  end


  # Idiomatic facade: client.item.list / client.item.load({ "id" => ... })
  def item
    require_relative 'entity/item_entity'
    @item ||= ItemEntity.new(self, nil)
  end

  # Deprecated: use client.item instead.
  def Item(data = nil)
    require_relative 'entity/item_entity'
    ItemEntity.new(self, data)
  end


  # Idiomatic facade: client.location.list / client.location.load({ "id" => ... })
  def location
    require_relative 'entity/location_entity'
    @location ||= LocationEntity.new(self, nil)
  end

  # Deprecated: use client.location instead.
  def Location(data = nil)
    require_relative 'entity/location_entity'
    LocationEntity.new(self, data)
  end


  # Idiomatic facade: client.monster.list / client.monster.load({ "id" => ... })
  def monster
    require_relative 'entity/monster_entity'
    @monster ||= MonsterEntity.new(self, nil)
  end

  # Deprecated: use client.monster instead.
  def Monster(data = nil)
    require_relative 'entity/monster_entity'
    MonsterEntity.new(self, data)
  end


  # Idiomatic facade: client.motion_value.list / client.motion_value.load({ "id" => ... })
  def motion_value
    require_relative 'entity/motion_value_entity'
    @motion_value ||= MotionValueEntity.new(self, nil)
  end

  # Deprecated: use client.motion_value instead.
  def MotionValue(data = nil)
    require_relative 'entity/motion_value_entity'
    MotionValueEntity.new(self, data)
  end


  # Idiomatic facade: client.skill.list / client.skill.load({ "id" => ... })
  def skill
    require_relative 'entity/skill_entity'
    @skill ||= SkillEntity.new(self, nil)
  end

  # Deprecated: use client.skill instead.
  def Skill(data = nil)
    require_relative 'entity/skill_entity'
    SkillEntity.new(self, data)
  end


  # Idiomatic facade: client.weapon.list / client.weapon.load({ "id" => ... })
  def weapon
    require_relative 'entity/weapon_entity'
    @weapon ||= WeaponEntity.new(self, nil)
  end

  # Deprecated: use client.weapon instead.
  def Weapon(data = nil)
    require_relative 'entity/weapon_entity'
    WeaponEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = MonsterHunterWorldSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/monsterhunterworld_sdk.py
# MonsterHunterWorld SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import MonsterHunterWorldUtility
from core.spec import MonsterHunterWorldSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import MonsterHunterWorldBaseFeature
from features import _make_feature


class MonsterHunterWorldSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = MonsterHunterWorldUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return MonsterHunterWorldUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = MonsterHunterWorldSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def ailment(self):
        """Idiomatic facade: client.ailment.list() / client.ailment.load({"id": ...})."""
        from entity.ailment_entity import AilmentEntity
        cached = getattr(self, "_ailment", None)
        if cached is None:
            cached = AilmentEntity(self, None)
            self._ailment = cached
        return cached

    def Ailment(self, data=None):
        # Deprecated: use client.ailment instead.
        from entity.ailment_entity import AilmentEntity
        return AilmentEntity(self, data)


    @property
    def armor(self):
        """Idiomatic facade: client.armor.list() / client.armor.load({"id": ...})."""
        from entity.armor_entity import ArmorEntity
        cached = getattr(self, "_armor", None)
        if cached is None:
            cached = ArmorEntity(self, None)
            self._armor = cached
        return cached

    def Armor(self, data=None):
        # Deprecated: use client.armor instead.
        from entity.armor_entity import ArmorEntity
        return ArmorEntity(self, data)


    @property
    def armor_set(self):
        """Idiomatic facade: client.armor_set.list() / client.armor_set.load({"id": ...})."""
        from entity.armor_set_entity import ArmorSetEntity
        cached = getattr(self, "_armor_set", None)
        if cached is None:
            cached = ArmorSetEntity(self, None)
            self._armor_set = cached
        return cached

    def ArmorSet(self, data=None):
        # Deprecated: use client.armor_set instead.
        from entity.armor_set_entity import ArmorSetEntity
        return ArmorSetEntity(self, data)


    @property
    def charm(self):
        """Idiomatic facade: client.charm.list() / client.charm.load({"id": ...})."""
        from entity.charm_entity import CharmEntity
        cached = getattr(self, "_charm", None)
        if cached is None:
            cached = CharmEntity(self, None)
            self._charm = cached
        return cached

    def Charm(self, data=None):
        # Deprecated: use client.charm instead.
        from entity.charm_entity import CharmEntity
        return CharmEntity(self, data)


    @property
    def decoration(self):
        """Idiomatic facade: client.decoration.list() / client.decoration.load({"id": ...})."""
        from entity.decoration_entity import DecorationEntity
        cached = getattr(self, "_decoration", None)
        if cached is None:
            cached = DecorationEntity(self, None)
            self._decoration = cached
        return cached

    def Decoration(self, data=None):
        # Deprecated: use client.decoration instead.
        from entity.decoration_entity import DecorationEntity
        return DecorationEntity(self, data)


    @property
    def event(self):
        """Idiomatic facade: client.event.list() / client.event.load({"id": ...})."""
        from entity.event_entity import EventEntity
        cached = getattr(self, "_event", None)
        if cached is None:
            cached = EventEntity(self, None)
            self._event = cached
        return cached

    def Event(self, data=None):
        # Deprecated: use client.event instead.
        from entity.event_entity import EventEntity
        return EventEntity(self, data)


    @property
    def item(self):
        """Idiomatic facade: client.item.list() / client.item.load({"id": ...})."""
        from entity.item_entity import ItemEntity
        cached = getattr(self, "_item", None)
        if cached is None:
            cached = ItemEntity(self, None)
            self._item = cached
        return cached

    def Item(self, data=None):
        # Deprecated: use client.item instead.
        from entity.item_entity import ItemEntity
        return ItemEntity(self, data)


    @property
    def location(self):
        """Idiomatic facade: client.location.list() / client.location.load({"id": ...})."""
        from entity.location_entity import LocationEntity
        cached = getattr(self, "_location", None)
        if cached is None:
            cached = LocationEntity(self, None)
            self._location = cached
        return cached

    def Location(self, data=None):
        # Deprecated: use client.location instead.
        from entity.location_entity import LocationEntity
        return LocationEntity(self, data)


    @property
    def monster(self):
        """Idiomatic facade: client.monster.list() / client.monster.load({"id": ...})."""
        from entity.monster_entity import MonsterEntity
        cached = getattr(self, "_monster", None)
        if cached is None:
            cached = MonsterEntity(self, None)
            self._monster = cached
        return cached

    def Monster(self, data=None):
        # Deprecated: use client.monster instead.
        from entity.monster_entity import MonsterEntity
        return MonsterEntity(self, data)


    @property
    def motion_value(self):
        """Idiomatic facade: client.motion_value.list() / client.motion_value.load({"id": ...})."""
        from entity.motion_value_entity import MotionValueEntity
        cached = getattr(self, "_motion_value", None)
        if cached is None:
            cached = MotionValueEntity(self, None)
            self._motion_value = cached
        return cached

    def MotionValue(self, data=None):
        # Deprecated: use client.motion_value instead.
        from entity.motion_value_entity import MotionValueEntity
        return MotionValueEntity(self, data)


    @property
    def skill(self):
        """Idiomatic facade: client.skill.list() / client.skill.load({"id": ...})."""
        from entity.skill_entity import SkillEntity
        cached = getattr(self, "_skill", None)
        if cached is None:
            cached = SkillEntity(self, None)
            self._skill = cached
        return cached

    def Skill(self, data=None):
        # Deprecated: use client.skill instead.
        from entity.skill_entity import SkillEntity
        return SkillEntity(self, data)


    @property
    def weapon(self):
        """Idiomatic facade: client.weapon.list() / client.weapon.load({"id": ...})."""
        from entity.weapon_entity import WeaponEntity
        cached = getattr(self, "_weapon", None)
        if cached is None:
            cached = WeaponEntity(self, None)
            self._weapon = cached
        return cached

    def Weapon(self, data=None):
        # Deprecated: use client.weapon instead.
        from entity.weapon_entity import WeaponEntity
        return WeaponEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
